internal/stanza/converter.go

```diff
@@ -15,49 +15,294 @@
+		c.wg.Wait()
+		close(c.pLogsChan)
+	})
+}
+
+// Channel returns the channel on which converted entries will be sent to.
+func (c *Converter) OutChannel() <-chan pdata.Logs {
+	return c.pLogsChan
+}
+
+// flush flushes all the accumulated pdata.Logs entries onto a channel.
+func (c *Converter) flush(ctx context.Context, pLogs []pdata.Logs) error {
+	doneChan := ctx.Done()
+
+	for _, pLog := range pLogs {
+		select {
+		case <-doneChan:
+			return fmt.Errorf("flushing log entries interrupted, err: %v", ctx.Err())
+
+		case c.pLogsChan <- pLog:
+
+		// The converter has been stopped so bail the flush.
+		case <-c.stopChan:
+			return nil
+		}
+	}
+
+	return nil
+}
+
+// Batch takes in an entry.Entry and aggregates it with other entries
+// that came from the same Resource.
+// If the maxFlushCount has been reached then trigger a flush via the flushChan.
+func (c *Converter) Batch(e *entry.Entry) error {
+	b, err := json.Marshal(e.Resource)
+	if err != nil {
+		return err
+	}
+
+	resourceHash := string(b)
+
+	c.dataMutex.Lock()
+
+	resourceEntries, ok := c.data[resourceHash]
+	if !ok {
+		// If we don't have any log entries for this Resource then create
+		// the provider entry in the cache for it.
+		resourceEntries = make([]*entry.Entry, 0, 8)
+	}
+
+	c.data[resourceHash] = append(resourceEntries, e)
+	c.dataCount++
+
+	needToFlush := c.dataCount >= c.maxFlushCount
+	c.dataMutex.Unlock()
+
+	// Flush max size has been reached: schedule a log flush.
+	if needToFlush {
+		pLogs := c.convertBuffer()
+		go c.queueForFlush(pLogs)
+	}
+
+	return nil
+}
+
+// convertBuffer converts the accumulated entries in the buffer and empties it.
+func (c *Converter) convertBuffer() []pdata.Logs {
+	c.dataMutex.Lock()
+	pLogs := make([]pdata.Logs, 0, len(c.data))
+	for h, entries := range c.data {
+		pLogs = append(pLogs, convertEntries(entries))
+		delete(c.data, h)
+	}
+	c.dataCount = 0
+	c.dataMutex.Unlock()
+
+	return pLogs
+}
+
+// queueForFlush queues the provided slice of pdata.Logs for flushing.
+func (c *Converter) queueForFlush(pLogs []pdata.Logs) {
+	select {
+	case c.flushChan <- pLogs:
+	case <-c.stopChan:
+	}
+}
+
+// convertEntries converts takes in a slice of entries coming from the same
+// Resource and converts them into a pdata.Logs.
+func convertEntries(entries []*entry.Entry) pdata.Logs {
 	out := pdata.NewLogs()
+	if len(entries) == 0 {
+		return out
+	}
+
 	logs := out.ResourceLogs()
 	logs.Resize(1)
 	rls := logs.At(0)
 
-	resource := rls.Resource()
-	if len(obsLog.Resource) > 0 {
+	// NOTE: This assumes that passed in entries all come from the same Resource.
+	if len(entries[0].Resource) > 0 {
+		resource := rls.Resource()
 		resourceAtts := resource.Attributes()
-		for k, v := range obsLog.Resource {
+		for k, v := range entries[0].Resource {
 			resourceAtts.InsertString(k, v)
 		}
 	}
 
 	rls.InstrumentationLibraryLogs().Resize(1)
 	ills := rls.InstrumentationLibraryLogs().At(0)
 
+	for i := 0; i < len(entries); i++ {
```

Comment:
Nit: preallocate ills.Logs() to have the size equal to len(entries) using `Resize` and then use `At` instead of `Append`. See `Resize` docs. This results in less reallocations of the slice.